Bottom: 27c3f8f8bfcb029109dfffdc2b681a4bc7ba0c0d
Top:    1a047d60534cab2efb0af8f927d0efcf1783c881
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-02-17 21:23:46 +0800



---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 7df78a53..e90c5fe 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -378,7 +378,8 @@ static bool virtnet_xdp_xmit(struct virtnet_info *vi,
 		return false;
 	}
 
-	virtqueue_kick(sq->vq);
+	if (virtqueue_kick_prepare(sq->vq))
+		virtqueue_notify(sq->vq);
 	return true;
 }
 
diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 2fe3535..b167f2c 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -391,6 +391,8 @@ static void handle_tx(struct vhost_net *net)
 	if (!vq_iotlb_prefetch(vq))
 		goto out;
 
+	printk("start tx!\n");
+
 	vhost_disable_notify(&net->dev, vq);
 
 	hdr_size = nvq->vhost_hlen;
@@ -411,8 +413,10 @@ static void handle_tx(struct vhost_net *net)
 						ARRAY_SIZE(vq->iov),
 						&out, &in);
 		/* On error, stop handling until the next kick. */
-		if (unlikely(head < 0))
+		if (unlikely(head < 0)) {
+			printk("head < 0\n");
 			break;
+		}
 		/* Nothing new?  Wait for eventfd to tell us they refilled. */
 		if (head == vq->num) {
 			if (unlikely(vhost_enable_notify(&net->dev, vq))) {
@@ -422,6 +426,7 @@ static void handle_tx(struct vhost_net *net)
 			break;
 		}
 		if (in) {
+			printk("unexpected format !\n");
 			vq_err(vq, "Unexpected descriptor format for TX: "
 			       "out %d, int %d\n", out, in);
 			break;
@@ -432,6 +437,7 @@ static void handle_tx(struct vhost_net *net)
 		iov_iter_advance(&msg.msg_iter, hdr_size);
 		/* Sanity check */
 		if (!msg_data_left(&msg)) {
+			printk("header unexpected!\n");
 			vq_err(vq, "Unexpected header len for TX: "
 			       "%zd expected %zd\n",
 			       len, hdr_size);
@@ -476,6 +482,7 @@ static void handle_tx(struct vhost_net *net)
 		/* TODO: Check specific error and bomb out unless ENOBUFS? */
 		err = sock->ops->sendmsg(sock, &msg, len);
 		if (unlikely(err < 0)) {
+			printk("sendmsg error!\n");
 			if (zcopy_used) {
 				vhost_net_ubuf_put(ubufs);
 				nvq->upend_idx = ((unsigned)nvq->upend_idx - 1)
